File: Python/Medium/graphs/cityWithSmallestNumberofNeighbors.py
import collections
def findTheCity(self, n, edges, distanceThreshold):
    graph = collections.defaultdict(dict)
    for n1, n2, distance in edges:
        graph[n1][n2] = distance
        graph[n2][n1] = distance

    def BFS(node):
        distances = [float('inf') if x != node else 0 for x in range(n)]
        queue = [node]

        while queue:
            cur = queue.pop(0)
            for neighbor,direct_distance in graph[cur].items():
                new_distance = direct_distance + distances[cur]
                if new_distance < distances[neighbor]:
                    distances[neighbor] = new_distance
                    queue.append(neighbor)

        return distances
    
    all_cities = [BFS(i) for i in range(n)]

    min_count = float('inf')
    result = -1

    for node, distance_arr in enumerate(all_cities):
        count = sum(1 for d in distance_arr if d <= distanceThreshold and d > 0)
        if count <= min_count:
            min_count = count
            result = node

    return result

File: JavaScript/Medium/graphs/cityWithSmallestNumberOfNeighbors.js
// Dijkstra's Algorithm

var findTheCity = function(n, edges, distanceThreshold){
    let graph = {}
    for (const [n1,n2, distance] of edges) {
        if (!graph[n1]) graph[n1] = {}
        if (!graph[n2]) graph[n2] = {}
        graph[n1][n2] = distance
        graph[n2][n1] = distance
    }

    var BFS = (i) => {
        let distances = new Array(n).fill(Infinity)
        distances[i] = 0
        let queue = [i]

        while (queue.length) {
            cur = queue.shift()
            if (graph[cur]) {
                for (const [neighbor, directDistance] of Object.entries(graph)) {
                    let newDistance = distances[cur] + directDistance
                    if (newDistance < distances[neighbor]) {
                        distances[neighbor] = newDistance
                        queue.push(neighbor)
                    }
                }
            }
        }

        return distances
    }

    let results = []
    for (let i = 0; i<n; i++) {
        results.push(BFS(i))
    }

    let minCount = Infinity
    let result = -1

    results.forEach((value, index) => {
        let count = value.filter(d => d<= distanceThreshold && d >0).length
        if (count <= minCount){
            minCount = count
            result = index
        }
    })

    return result

}